FAI-105: Embed kogito-tooling editors in trusty-ai

manstis opened the pull request:
See https://issues.redhat.com/browse/FAI-105

Hi @tiagobento @ederign this is my first pass at making a generic embeddable _editor_ for kogito-tooling. Most of it will look very familiar to you (as I see the envelope etc is copy and pasted across the different channels already). `index.tsx` (the non-envelope version) and `KogitoEditorWrapper` are the real points of interest to demonstrate a generic embedded use case.

I do have a problem with a forward reference to the _real_ editor to get it's content (if you run this you'll see there are two buttons to the content; one is _external_ to the embedded editor the other is _internal_. The _external_ form does not run anything.. IDK why, any ideas?)

All existing channels use the _internal_ form to get the editors' content and perform a save.

Trusty does not need to get the content and would set the `readonly` flag always to true. Therefore, for my specific use-case, the _external_ forward ref to get the content does not cause me a problem. However if you want to consider a generic ability to embed editors it'll need some work (and to expose the remaining envelope API methods).

WDYT?

@kelvah FYI.

manstis commented on packages/embedded-editor/src/editor/Editor.tsx:
```diff
@@ -0,0 +1,115 @@
+      },
+      receive_resourceContentRequest(resourceContentRequest: ResourceContentRequest) {
+        console.debug(`Resource Content Request`);
+        self.respond_resourceContent(new ResourceContent(resourceContentRequest.path, undefined));
```
Optional functions to handle the Resource API can be provides as properties to `KogitoEditorWrapper`

manstis commented on packages/embedded-editor/src/editor/Editor.tsx:
```diff
@@ -0,0 +1,115 @@
+        console.debug(`Resource List Request`);
+        self.respond_resourceList(new ResourcesList(request.pattern, []));
+      },
+      notify_editorUndo: (edits: ReadonlyArray<KogitoEdit>) => {
```
Likewise for the State API.

manstis commented:
It's plausible this _embedded_ form could be re-used in all the existing channels; rather than repeat the envelope magic, once the necessary API handler functions are passed to `Editor` as properties from `KogitoEditorWrapper`.

tiagobento commented:
Hi @manstis! Please tell me if I'm correct on what I understood from your PR.

1. The intent of this draft PR is to showcase how an "embedded editor" component would look like.
2. `KogitoEditorWrapper.tsx` is a first version of what could be a good candidate for such component.
3. `Editor.tsx` is essentially a copy of what's on the other channels.
4. The webapp on this new `embedded-editor` package is for demonstration purposes only and would not evolve to be a "deliverable" webapp.

If I actually understood everything correctly, I have some observations, but I'll wait for your response :)

manstis commented:
@tiagobento Your observations are correct on all accounts.

tiagobento commented:
@manstis Cool! So here's what I think:

1. Having an `envelope/index.js` and an `envelope/index.html` is necessary. There's no way we can make that reusable if you don't want to depend on external services (which I agree is a good decision). Also, you can potentially have different Routes, if you support different editors.
2. Pretty much everything that's written on `Editor.tsx` is channel-specific. Every channel has its own way of registering the communication with the envelope, its own iframe CSS needs, it's own actions to do when receiving a message etc. I agree that it doesn't _look_ like a component atm because you have to use bits of code that are not necessarily related to it, but the abstraction is there. What I think could be a good option is to extract some props, inverting the control, so that the parent of `Editor.tsx` can decide what to do on each message received, how to register it etc. You don't have to use a context on this component too, since using context was only done on other channels to make the code cleaner.
3. Having this componentized version of `Editor.tsx` would allow you to create a `ReadonlyEditor.tsx` on top of if, since it's a specification of the complete API. WDYT?
4. I don't understand what "internal" and "external" means on your comments, can you provide more info about that?

manstis commented:
Hi @tiagobento I've made some changes following the points you mention above.

It might be easier for us to meet and discuss them rather than exchange messages.

manstis commented:
Hi @tiagobento I've changed the `online-editor` to use the embedded form.

Perhaps we can meet again to discuss changes and limitations.

manstis commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -128,7 +105,7 @@ export function HomePage(props: Props) {
   const onFileUploadFromDnd = useCallback((file: File) => {
     const fileExtension = extractFileExtension(file.name);
-    if (!fileExtension || !context.router.getLanguageData(fileExtension)) {
+    if (!fileExtension || !isKnownEditorType(fileExtension)) {
```
I might add back use of the router to check the file type is supported... this was changed during my hack-fest!

manstis replied:
Reinstated use of `router`.

manstis commented on packages/online-editor/src/envelope/envelope.ts:
```diff
@@ -14,10 +14,10 @@
  * limitations under the License.
  */
 
-import * as MicroEditorEnvelope from "@kogito-tooling/microeditor-envelope";
```
I tried to remove the need for `envelope.ts` from the _consuming_ application however failed.

Ideally this should be in the `embedded-editor` module. However the `embedded-editor` module needs to exclude React in its build/packaging otherwise the "Invalid hook call" [error](https://reactjs.org/warnings/invalid-hook-call-warning.html) error is thrown - as both the _embedded_ and _online_ modules import React leading to two instances.

Excluding React however leads to a module loader error in the `IFRAME` as one of `envelope.js` dependencies tries to load React-DOM however this fails as React was not loaded due to it being excluded. So.. in short `envelope.js` needs to be in the _consuming_ application unless we can find another solution.

manstis replied:
This has successfully been moved back to `embedded-editor` thanks to a nifty bit of webpack configuration.

manstis commented on packages/online-editor/webpack.config.js:
```diff
@@ -70,12 +70,12 @@ module.exports = async (env, argv) => {
         { from: "./static/index.html", to: "./index.html" },
         { from: "./static/favicon.ico", to: "./favicon.ico" },
-        { from: "../kie-bc-editors-unpacked", to: "./gwt-editors" }
+        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/envelope", to: "./envelope" },
```
I have to copy the gwt-editors to the _consuming_ webapp... they're bundled in `embedded_editor`. 

Ideally I'd like to have been able to reference them from the bundle I'm not sure it's possible.

It might be better to exclude them from the `embedded-editor` bundle and just copy them from the unpacked folder?

manstis commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -24,6 +24,7 @@ import { EmbeddedEditorRouter } from "./EmbeddedEditorRouter";
 interface Props {
   file: File;
   router: EmbeddedEditorRouter;
+  channelType: ChannelType;
```
Allows us to configure the `ChannelType` where the `embedded-editor` is re-used.

manstis commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -34,6 +35,7 @@ interface Props {
   onEditorRedo?: (edits: ReadonlyArray<KogitoEdit>) => void;
   onNewEdit?: (edit: KogitoEdit) => void;
   onPreviewResponse?: (previewSvg: string) => void;
+  envelopeURI?: string;
```
Allows us to override the default `envelope.html` with a custom one, if needed.

manstis commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -107,6 +109,8 @@ const RefForwardingEmbeddedEditor: React.RefForwardingComponent<EmbeddedEditorRe
     }
   }, [props.onPreviewResponse]);
 
+  const envelopeURI = props.envelopeURI ? props.envelopeURI : "envelope/envelope.html";
```
Fallback to default it custom is not provided.

manstis commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -189,8 +193,9 @@ const RefForwardingEmbeddedEditor: React.RefForwardingComponent<EmbeddedEditorRe
-        src={"envelope/envelope.html"}
+        src={envelopeURI}
         title="Kogito editor">
+        <div id="kogito--editor--channel--type">{props.channelType}</div>
```
Pass `ChannelType` to `IFRAME`....

manstis commented on packages/embedded-editor/src/envelope/envelope.ts:
```diff
@@ -30,5 +30,9 @@ MicroEditorEnvelope.init({
   },
   editorFactory: new GwtEditorWrapperFactory(gwtAppFormerApi, new DefaultXmlFormatter()),
-  editorContext: { channel: ChannelType.ONLINE }
+  editorContext: { channel: getChannelType() }
```
... where I can then read it to setup the context correctly.

manstis commented:
@tiagobento This PR is now complete (other than tests for `embedded-editor`).

Both `online` and `desktop` re-use the editor and default envelope in `embedded-editor`.

Did you want to review (with @paulovmr?) before I finish writing tests?

manstis commented on packages/desktop/src/webview/editor/EditorPage.tsx:
```diff
@@ -226,6 +226,13 @@ export function EditorPage(props: Props) {
     };
   }, []);
 
+  const file: EmbeddedFile = {
+    fileName: context.file?.filePath || "",
+    editorType: context.file?.fileType as EditorType,
+    getFileContents: () => Promise.resolve(context.file?.fileContent || ""),
+    isReadOnly: false
+  };
```
Updated. Thanks for the pointer to the _nullish_ operator.

manstis commented on packages/embedded-editor/src/common/File.ts:
```diff
@@ -0,0 +1,47 @@
+ * limitations under the License.
+ */
+
+export type EditorType = "dmn" | "bpmn";
```
Done.

manstis commented on packages/embedded-editor/webpack.config.js:
```diff
@@ -0,0 +1,101 @@
+    plugins: [
+      new CopyPlugin([
+        { from: "./static/envelope", to: "./envelope" },
+        { from: "../kie-bc-editors-unpacked/dmn", to: "./gwt-editors/dmn" },
+        { from: "../kie-bc-editors-unpacked/bpmn", to: "./gwt-editors/bpmn" }
```
Done.

manstis commented on packages/embedded-editor/package.json:
```diff
@@ -0,0 +1,36 @@
+    "@kogito-tooling/kie-bc-editors": "0.4.2",
+    "@kogito-tooling/microeditor-envelope": "0.4.2",
+    "@kogito-tooling/microeditor-envelope-protocol": "0.4.2",
+    "@octokit/rest": "16.35.0"
```
It's not used. Removed.

manstis commented on packages/desktop/package.json:
```diff
@@ -14,6 +14,7 @@
     "@kogito-tooling/kie-bc-editors": "0.4.2",
     "@kogito-tooling/microeditor-envelope": "0.4.2",
     "@kogito-tooling/microeditor-envelope-protocol": "0.4.2",
```
Removed (and similarly for `online`).

manstis commented:
@tiagobento I've updated the PR to pass `dataenvelopechannel` as an additional attribute on the `IFRAME`.. I was not able to call it `data-envelope-channel` nor `dataEnvelopeChannel` as TS complained custom properties needed to be all lowercase. I still think it's better (as an additional attribute than using a `DIV` to pass the parameter). WDYT?

manstis commented:
Changes to `chrome-extension` to follow in the next commit...

manstis commented on packages/desktop/webpack.config.js:
```diff
@@ -92,7 +92,8 @@ module.exports = [
         { from: "./static/index.html", to: "./index.html" },
         { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/envelope", to: "./envelope" },
-        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/gwt-editors", to: "./gwt-editors" }
+        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/gwt-editors/dmn", to: "./gwt-editors/dmn" },
```
I only copy `DMN` and `BPMN` as `SCESIM` is not supported in the `Desktop` channel at the moment.

manstis commented on packages/online-editor/webpack.config.js:
```diff
@@ -74,7 +74,8 @@ module.exports = async (env, argv) => {
         { from: "./static/favicon.ico", to: "./favicon.ico" },
         { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/envelope", to: "./envelope" },
-        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/gwt-editors", to: "./gwt-editors" }
+        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/gwt-editors/dmn", to: "./gwt-editors/dmn" },
```
I only copy `DMN` and `BPMN` as `SCESIM` is not supported in the `online` channel at the moment.

manstis commented:
@tiagobento All done. The Chrome extension has also been ported to re-use `EmbeddedEditor`.

I gave the _new_ Chrome extension a quick try out and it seems to work fine still (phew!).. but it'd probably be a very good idea for somebody with greater knowledge of its capabilities to give it a good kicking too.. The test coverage is rather light. Obviously we cannot have any of the pre-existing channels to break as it seems interest in `kogito-tooling` is quite good at the moment :-)

tiagobento commented:
@manstis Regarding `data-envelope-channel`, it _should_ work. See https://www.typescriptlang.org/docs/handbook/jsx.html#attribute-type-checking

Can you paste the error you encountered?

tiagobento commented on packages/chrome-extension/resources/style.css:
```diff
@@ -199,3 +199,18 @@ b.icon.cross:before {
   content: "❌";
   font-size: 0.6em;
 }
+
+.kogito--editor--container {
+  display: flex; 
+  width: 100%; 
+  height: 100%; 
+  flex-direction: column; 
+  overflow: hidden;
+}
+
+.kogito--editor--container iframe {
+  flex-grow: 1; 
+  border: none; 
+  margin: 0; 
+  padding: 0;
+}
```
Looks like this bit of code is "duplicated" (actually it shows up more than two times but I never know what term to use. "replicated" maybe? 😄) across other modules. It would be nice if this could be contained on `embedded-editor` only.

manstis replied:
Moved to inline CSS in `EmbeddedEditor`.. I preferred to inline the CSS than have a shared CSS file there as I'd need to ensure the CSS file was added to each channels main HTML page.

tiagobento commented on packages/chrome-extension/src/app/components/common/KogitoEditorIframe.tsx:
```diff
@@ -34,98 +34,28 @@ interface Props {
-      })
-    );
-  }, [router, resourceContentService]);
+  //Wrap file content into object for EmbeddedEditor
+  const file: File = {
+    fileName: props.contentPath,
+    editorType: props.openFileExtension as EditorType,
+    getFileContents: props.getFileContents,
+    isReadOnly: props.readonly
+  }
```
This is where React Hooks get a little tricky. This bit of code is inside a function component, so React will execute this *entire* function every time it needs to *check* if there's something to render. To avoid re-creating objects every time this function executes, React has a memoization hook -- `useMemo`. For functions, React has `useCallback`. As a general rule, every local variable on a function component should be created using a hook. In React (by convention), hooks are always functions prefixed by `use`. Because of that, as a future hint, we should never write functions starting with `use` if they're not a React Hook.

tiagobento replied:
Also, just as other other hooks, `useMemo` and `useCallback` have dependencies. This is how you tell React that this memoized value should be computed again. In this case, you should pass all the props to it. Note that it compares using references, not values, so you have to list everything there so that React can re-compute it when some of those 4 props change.

tiagobento replied:
It sounds like a very inoffensive object creation, but since React compares references, it'll think you have a different File reference every time this function executes, so child components depending on this `file` will be re-evaluated as well, creating a cascading effect, potentially resulting in poor performance or bugs.

manstis replied:
Done.

tiagobento commented on packages/chrome-extension/src/app/components/common/KogitoEditorIframe.tsx:
```diff
@@ -136,38 +66,42 @@ const RefForwardingKogitoEditorIframe: React.RefForwardingComponent<IsolatedEdit
+      return;
+    }
+
+    //keep line breaks
+    const content = editorContent.content.split("\n").join("\\n");
 
-    return () => {
-      envelopeBusOuterMessageHandler.stopInitPolling();
-      window.removeEventListener("message", listener);
-    };
-  }, [envelopeBusOuterMessageHandler]);
+    runScriptOnPage(
+      `document.querySelector("${GITHUB_CODEMIRROR_EDITOR_SELECTOR}").CodeMirror.setValue('${content}')`
+    );
+  };
```
Same here. Use `useCallback` to memoize this function.

tiagobento replied:
Don't forget the dependencies (if any)!

manstis replied:
Done.

tiagobento commented on packages/desktop/src/webview/editor/EditorPage.tsx:
```diff
@@ -226,6 +226,13 @@ export function EditorPage(props: Props) {
     };
   }, []);
 
+  const file: EmbeddedFile = {
+    fileName: context.file?.filePath ?? "",
+    editorType: context.file?.fileType as EditorType,
+    getFileContents: () => Promise.resolve(context.file?.fileContent ?? ""),
+    isReadOnly: false
+  };
```
Memoize that.

manstis replied:
Done.

tiagobento commented on packages/desktop/src/webview/home/FilesPage.tsx:
```diff
@@ -72,7 +47,7 @@ enum ImportFileErrorType {
 const ALERT_AUTO_CLOSE_TIMEOUT = 3000;
 
-const typeFilterOptions = [{ value: "All" }, { value: "BPMN" }, { value: "DMN" }];
+const typeFilterOptions = [{ value: "All" }, { value: EditorType.BPMN.toUpperCase() }, { value: EditorType.DMN.toUpperCase() }];
```
Hmm..... I'm not sure you should be using `EditorType` here. These values are for display, not for logic, `"BPMN"` and `EditorType.BPMN.toUpperCase()` are two distinct things despite their equal value. Same for DMN.

manstis replied:
Reverted back.

tiagobento commented on packages/embedded-editor/src/embedded/AugmentedIFrame.ts:
```diff
@@ -13,23 +13,14 @@
 
-import { Router, Routes } from "@kogito-tooling/core-api";
+declare module 'react' {
 
-export class OnlineEditorRouter extends Router {
-  constructor(...routesArray: Routes[]) {
-    super(...routesArray);
-  }
+    // Add the additonal attribute `dataenvelopechannel` to IFRAMES.
+    interface IframeHTMLAttributes<T> extends React.HTMLAttributes<T> {
+        dataenvelopechannel: ChannelType;
+    }
```
Yeah.. this entire file should not be necessary. There must be some configuration we're missing. Probably on tsconfig or tslint.

manstis replied:
OK, so I see now... "... If an attribute name is not a valid JS identifier (like a `data-*` attribute), it is not considered to be an error if it is not found in the element attributes type..." and this typed extension to `IFRAME` attributes is not needed. Removed and changed to use `data-envelope-channel` as you originally proposed!

tiagobento commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -0,0 +1,215 @@
+    if (props.onSetContentError) {
+      props.onSetContentError();
+    }
+  }, [props.onContentResponse]);
```
Oops 😝

manstis replied:
Corrected.

tiagobento commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -0,0 +1,215 @@
+    }
+  }, [props.onPreviewResponse]);
+
+  const envelopeUri = props.envelopeUri ?? "envelope/envelope.html";
```
Memoize this.

manstis replied:
Fixed.

tiagobento commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -0,0 +1,215 @@
+          onPreviewRequest(previewSvg);
+        }
+      }));
+  }, [props]);
```
Can you double-check this dependencies list?

manstis replied:
I've changed his to list the properties accessed by the memoised `EnvelopeBusOuterMessageHandler` implementation. I had thought an equality check was performed on the dependency (hence I simply used `props` previously).

tiagobento commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -0,0 +1,215 @@
+  );
+
+  return (
+    <div className="kogito--editor--container">
+      <iframe
+        ref={iframeRef}
+        id={"kogito-iframe"}
+        src={envelopeUri}
+        title="Kogito editor"
+        dataenvelopechannel={props.channelType} />
+    </div>
```
Do we need a containing `div`? I figure you could apply the same styles to the `iframe` directly?

manstis replied:
Seems not, so I've removed.

tiagobento commented on packages/embedded-editor/src/embedded/EmbeddedViewer.tsx:
```diff
@@ -0,0 +1,67 @@
+  }, [props.onResourceListRequest]);
+
+  return (
+    <EmbeddedEditor
+      file={props.file}
+      router={props.router}
+      channelType={props.channelType}
+      onContentResponse={() => {/*NOP*/ }}
+      onSetContentError={() => {/*NOP*/ }}
+      onDirtyIndicatorChange={(isDirty: boolean) => {/*NOP*/ }}
+      onReady={() => {/*NOP*/ }}
+      onResourceContentRequest={onResourceContentRequest}
+      onResourceListRequest={onResourceListRequest}
+      onEditorUndo={(edits: ReadonlyArray<KogitoEdit>) => {/*NOP*/ }}
+      onEditorRedo={(edits: ReadonlyArray<KogitoEdit>) => {/*NOP*/ }}
+      onNewEdit={(edit: KogitoEdit) => {/*NOP*/ }}
+      onPreviewResponse={(previewSvg: string) => {/*NOP*/ }}
+      envelopeUri={props.envelopeUri}
+    />
```
All no-op functions should be memoized as well. `() => {}` is creating a new lambda, therefore creating a new reference, so React will think they changed, and will execute `EmbeddedEditor` etc etc.. You can create an empty `useCallback` and pass the same for all of them.

tiagobento replied:
Even for the one with parameters, I *think* you can do `const noop = useCallback((...args: any) => { /*NO OP*/ } , []);`

manstis replied:
Seems to work fine; so implemented as suggested.

tiagobento commented on packages/embedded-editor/src/envelope/envelope.ts:
```diff
@@ -30,5 +29,9 @@ MicroEditorEnvelope.init({
 });
+
+function getChannelType(): ChannelType {
+  return frameElement.attributes.getNamedItem("dataenvelopechannel")?.value as ChannelType;
```
Nice optional chaining 😍

tiagobento commented on packages/online-editor/src/index.tsx:
```diff
@@ -91,12 +94,13 @@ function openFileByUrl() {
 function openFile(filePath: string, getFileContent: Promise<string>) {
   const file = {
+    isReadOnly: false,
+    editorType: extractFileExtension(removeDirectories(filePath)!)! as EditorType,
     fileName: removeFileExtension(removeDirectories(filePath)!)!,
```
I know you just replicated line 99, but while you're at it, can you add some improved checks to get rid of these ! ! ! ! ?

manstis replied:
Improved (I think), to use nullify operator (you'll wish you never showed me it!!!)

tiagobento commented on packages/embedded-editor/webpack.config.js:
```diff
@@ -0,0 +1,102 @@
+        { from: "../kie-bc-editors-unpacked/scesim", to: "./gwt-editors/scesim" }
+      ])
+    ],
+    devServer: {
+      historyApiFallback: {
+        disableDotRule: true
+      },
+      disableHostCheck: true,
+      watchContentBase: true,
+      contentBase: [
+        path.join(__dirname, "./dist"),
+        path.join(__dirname, "./static")
+      ],
+      compress: true,
+      port: 9001
+    }
```
I guess this should be on the other object? I don't think it matters much, but idk, `index.ts` in this case is only for exporting purposes, right? :)

manstis replied:
Deleted.

manstis replied:
I also remove the `start` script from `package.json` to launch wds.

manstis commented:
@tiagobento OK, all done... for now :-)

I'm happy with incremental reviews.

manstis commented:
@caponetto I've added you too since I want to be sure your changes made [here](https://github.com/kiegroup/kogito-tooling/pull/145) are incorporated correctly. ALL channels (other than VSCode) i.e. Desktop, Online, Github, now share a common _editor_ in `embedded-editor`. AFAIK your referenced PR just changed the definition of the undo/redo envelope handler/messages. I've incorporated that into this PR [here](https://github.com/kiegroup/kogito-tooling/pull/142/commits/29ccef0a69da1c7c30dbbd5e7854b8d4c36b2a59).

paulovmr commented:
@manstis I'll leave the review with @tiagobento :)

manstis commented:
@tiagobento Unit tests now included; so this is OK for review.

I'd like to have improved the coverage of the use of `EnvelopeBusOuterMessageHandler` in the `EmbeddedEditor` however since it's _internal_ (by design, to hide it's complexities from the average user) IDK if it's possible. Any ideas?

manstis commented:
@tiagobento OK.. all tests now complete too.. close to 100% coverage.

Just need your approval before it can be merged :-)

tiagobento commented on packages/embedded-editor/webpack.config.js:
```diff
@@ -0,0 +1,89 @@
+    plugins: [
+      new CopyPlugin([
+        { from: "./static/envelope", to: "./envelope" },
+        { from: "../kie-bc-editors-unpacked/dmn", to: "./gwt-editors/dmn" },
+        { from: "../kie-bc-editors-unpacked/bpmn", to: "./gwt-editors/bpmn" },
+        { from: "../kie-bc-editors-unpacked/scesim", to: "./gwt-editors/scesim" }
```
@manstis After some thought I figured we should not include the editors inside this package. Editors should be managed by whomever is using the embedded-editors, since every app might have different needs. If you want to distribute the editors using kogito-tooling, I recommend creating a separate package, or even reusing `kie-editors-unpacked`. Note that have to change the `publish_artifacts.yml` GitHub action to actually publish it to NPM.

manstis replied:
Removed from `embedded-editor` webpack config.

manstis replied:
Where's `publish_artifacts.yml` live?

tiagobento replied:
`.github/workflows/`

manstis replied:
OK, I won't worry about deploying the unpacked editors for now.. we may replicate the unpacking done in `kogito-tooling` in `trusty` for now.. I'm going for the route of least resistance at the moment.

tiagobento commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -228,25 +205,27 @@ export function HomePage(props: Props) {
 
   const createEmptyDmnFile = useCallback(() => {
-    createEmptyFile("dmn");
+    createEmptyFile(EditorType.DMN);
```
Is it possible to change the parameter of `createEmptyFile` type to EditorType?

manstis replied:
Done.

tiagobento commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -256,11 +235,11 @@ export function HomePage(props: Props) {
 
   const tryBpmnSample = useCallback(() => {
-    trySample("bpmn");
+    trySample(EditorType.BPMN);
```
Is it possible to change the parameter of `trySample` type to EditorType?

manstis replied:
Done.

tiagobento commented on packages/online-editor/webpack.config.js:
```diff
@@ -70,12 +69,13 @@ module.exports = async (env, argv) => {
         { from: "./static/favicon.ico", to: "./favicon.ico" },
-        { from: "../kie-bc-editors-unpacked", to: "./gwt-editors" }
+        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/envelope", to: "./envelope" },
+        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/gwt-editors/dmn", to: "./gwt-editors/dmn" },
+        { from: "../../node_modules/@kogito-tooling/embedded-editor/dist/gwt-editors/bpmn", to: "./gwt-editors/bpmn" }
       ])
```
(*) This.

manstis replied:
Changed to use `kie-bc-editors-unpacked` path as we're no longer packaging the editors in `embedded-editor` but also removed use of the enum as you proposed.

tiagobento commented on packages/online-editor/src/__tests__/testing_utils.tsx:
```diff
@@ -14,33 +14,29 @@
+    routes: new Routes(),
+    router: new EmbeddedEditorRouter(
       new GwtEditorRoutes({
         bpmnPath: "gwt-editors/bpmn",
         dmnPath: "gwt-editors/dmn",
         scesimPath: "gwt-editors/scesim"
```
(*) This pairs nicely with...

tiagobento commented on packages/online-editor/src/App.tsx:
```diff
@@ -46,18 +44,16 @@ interface Props {
+    new EmbeddedEditorRouter(
+      new GwtEditorRoutes(
+        {
+          dmnPath: "gwt-editors/" + EditorType.DMN,
+          bpmnPath: "gwt-editors/" + EditorType.BPMN,
+          scesimPath: "gwt-editors/" + EditorType.SCESIM
```
(*) This doesn't pair nicely with...

I think the grep test is broken when you concat that. We're using a logic type that represents a type of an editor to decide the name of a directory. I'm not a fan of this type of coupling. Can you revert that to use plain strings?

manstis replied:
Reverted to plain strings.

tiagobento commented on packages/desktop/src/webview/App.tsx:
```diff
@@ -48,15 +47,13 @@ export function App(props: Props) {
-          bpmnPath: "editors/bpmn",
-          dmnPath: "editors/dmn",
-          scesimPath: "editors/scesim"
+          dmnPath: "gwt-editors/" + EditorType.DMN,
+          bpmnPath: "gwt-editors/" + EditorType.BPMN,
+          scesimPath: "gwt-editors/" + EditorType.SCESIM
```
Same here.

manstis replied:
Done.

tiagobento commented on packages/embedded-editor/src/__tests__/embedded/EmbeddedEditor.test.tsx:
```diff
@@ -0,0 +1,315 @@
+                    />
+                );
+
+                editorRef.current?.setContent("");
```
What is expected after this is executed? Normally, we tend to render a component, execute an action and check that the DOM was updated accordingly. In this case, there's nothing to be updated in the DOM, but you should create a way to verify that calling `setContent` reached a point where you're comfortable saying that it works.

manstis replied:
Yeah, you caught me out.. I know this test doesn't test anything. Let me improve it.

manstis replied:
Updated to _expect_ a method on the outer bus to be called.

tiagobento commented on packages/embedded-editor/src/__tests__/embedded/EmbeddedEditor.test.tsx:
```diff
@@ -0,0 +1,315 @@
+                    />
+                );
+
+                editorRef.current?.requestContent();
```
Same here. What should happen after a `requestContent` call? I assume you would receive something on the `receive_contentResponse` counterpart?

manstis replied:
Same again. I am such a naughty developer.

manstis replied:
Updated to _expect_ a method on the outer bus to be called.

tiagobento commented on packages/embedded-editor/src/__tests__/embedded/EmbeddedEditor.test.tsx:
```diff
@@ -0,0 +1,315 @@
+                    />
+                );
+
+                expect(envelopeFactorySpy).toBeCalled();
```
I don't think `EmbeddedViewer` tests should know about the existence of `envelopeFactory`. IMHO, tests should use components as you use them in the code itself.

manstis replied:
Tidied up whole `EmbeddedViewer` test.

tiagobento commented on packages/embedded-editor/src/__tests__/embedded/EmbeddedEditor.test.tsx:
```diff
@@ -0,0 +1,315 @@
+                    />
+                );
+
+                const envelopeFactory: EnvelopeBusOuterMessageHandler = envelopeFactorySpy.mock.results[0].value;
+                envelopeFactory.impl.receive_setContentError("error");
```
You can simulate this exact same behaviour by posting a message to `window`. This will exclude you from the burden of having to know how it's implemented.

manstis replied:
Can you please point me to where I can get the definitions of different messages and how to post them to `window`? I'll update these tests to post messages as you propose (now I've moved creation of `EnvelopeBusOuterMessageHandler` back into `EmbeddedEditor` itself.

tiagobento replied:
See `EditorEnvelopeController.test.tsx`

manstis replied:
....and I failed miserably and think I'm going to need some help here.

I looked at `EditorEnvelopeController.test.tsx` and cloned most of it to `EmbeddedViewer.test.tsx` (I had to export some extra [bits](https://github.com/kiegroup/kogito-tooling/pull/142/commits/ffff5f193c5cd1747e61005e6e52c2817f18fe2e#diff-eb6a1703eb9fe90d9eaa421338705167) from `microeditor-envelope`). The `EmbeddedViewer::init` test (copied from `EditorEnvelopeController.test.tsx`) passes but the two for get resource/resource list fail. 

Also, I notice my _snapshots_ do not list the `EmbeddedViewer` as I'd expected but what looks like the (mocked) envelope? 

Any chance you could clone this PR and look at `EmbeddedViewer.test.tsx` and give me some help? Once one test works I'll do the others. I don't think I know quite enough about the inner and outer message handlers and how they communicate with one and other quite yet...

Sorry.

tiagobento replied:
Will do!

manstis replied:
@tiagobento Tests updated and work - and have ascertions!

tiagobento commented on packages/embedded-editor/src/embedded/EnvelopeBusOuterMessageHandlerFactory.tsx:
```diff
@@ -0,0 +1,88 @@
+      receive_languageRequest() {
+        self.respond_languageRequest(props.router.getLanguageData(props.file.editorType));
+      },
+      receive_contentResponse(content: EditorContent) {
+        onContentResponse(content);
+      },
+      receive_contentRequest() {
+        props.file
+          .getFileContents()
+          .then(c => self.respond_contentRequest({ content: c ?? "", path: props.file.fileName }));
+      },
+      receive_setContentError(errorMessage: string) {
+        onSetContentError(errorMessage);
+      },
+      receive_dirtyIndicatorChange(isDirty: boolean) {
+        onDirtyIndicatorChange(isDirty);
+      },
+      receive_ready() {
+        onReady();
+      },
+      receive_resourceContentRequest(request: ResourceContentRequest) {
+        onResourceContentRequest(request).then(r => self.respond_resourceContent(r!));
+      },
+      receive_resourceListRequest(request: ResourceListRequest) {
+        onResourceListRequest(request).then(r => self.respond_resourceList(r!));
+      },
+      notify_editorUndo: () => {
+        onEditorUndo();
+      },
+      notify_editorRedo: () => {
+        onEditorRedo();
+      },
+      receive_newEdit(edit: KogitoEdit) {
+        onNewEdit(edit);
+      },
+      receive_previewRequest(previewSvg: string) {
+        onPreviewResponse(previewSvg);
+      }
+    }));
+}
```
Is this indirection necessary? I feel like it's only abstracting the creation of the `impl` object along with the `busApi`. This could be inlined directly on EmbeddedEditor.

manstis replied:
Yes, I extracted it to facilitate testing of the _closed_ `EmbeddedEditor` component.

However you say below I can test the bus interaction by posting messages to `window.`

I'll try to refactor and ping you if I have problems.

manstis replied:
Ping! See above, about where/how to post messages :-)

tiagobento replied:
See above.

tiagobento commented on packages/embedded-editor/src/embedded/EmbeddedEditor.tsx:
```diff
@@ -0,0 +1,193 @@
+        onNewEdit,
+        onPreviewRequest)
+    },
+    [props.router, props.file.editorType, props.file.fileName,
+    props.onContentResponse, props.onSetContentError, props.onDirtyIndicatorChange,
+    props.onReady, props.onResourceContentRequest, props.onResourceListRequest, props.onEditorUndo,
+    props.onEditorRedo, props.onNewEdit, props.onPreviewResponse]
```
After you inline newEnvelopeBusOuterMessageHandler here, There will be no need to create one `useCallback` for each method, you can simply do `props.method?.(args)` on the `impl` object. Think about when we want to create a new method on the envelope API. All that just makes it harder.

manstis replied:
I get the intent; avoiding needing to add additional dependencies on the hook; but you've lost me on the use of `props.method?.(args)` can you please try explaining again? I tried adding that statement as the dependency to the `useMemo(..)` but it just moaned `method` is (obviously?) not a property.

tiagobento replied:
`props.method?.(args)` means: Call method `method` of `props` with parameters `args` if it exists. Pretty much an optional chaining for function calls.

manstis replied:
Sure, I understand that; but I get a `Property 'method' does not exist on type 'Props'.ts(2339)` error when I change this to `useMemo(()=>......, [props.method?.(args)]`. I thought that is what you intended?

tiagobento replied:
Yeah.... I think this comment was placed in the wrong place. I meant to do something like that, now that you've inlined the creation of the `impl` object.

```
...
receive_contentResponse(content: EditorContent) {
    props.onContentResponse?.(content);
},
...
```

manstis replied:
OK, I understand now. I kept the code for the `ResourceContentApi` callbacks as they need to return empty responses if no implementations are explicitly provided. I also removed the most of the dependencies from the `EnvelopeBusOuterMessageHandler` memo construction.

manstis replied:
Let me know if I've still managed to get it wrong :-)

manstis commented on packages/desktop/src/webview/editor/EditorPage.tsx:
```diff
@@ -226,6 +226,15 @@ export function EditorPage(props: Props) {
     };
   }, []);
 
+  const file = useMemo(() => {
+    return {
+      fileName: context.file?.filePath ?? "",
+      editorType: context.file?.fileType as EditorType,
+      getFileContents: () => Promise.resolve(context.file?.fileContent ?? ""),
+      isReadOnly: false
+    }
```
I tried removing the `return {` and closing `}` but it was invalid syntax.

Did you mean something else?

tiagobento replied:
Yeah.. because you're returning an object, you have to wrap the object in `()`

tiagobento replied:
Otherwise the compiler doesn't know if you're providing a function body or an object

manstis replied:
Sorry, I am being incredibly thick, what should I change the following to. I've tried numerous combinations of removing brackets and `return` but they all give syntax errors:
```
  const file = useMemo(() => {
    return {
      fileName: context.file?.filePath ?? "",
      editorType: context.file?.fileType as EditorType,
      getFileContents: () => Promise.resolve(context.file?.fileContent ?? ""),
      isReadOnly: false
    }
  }, [context.file?.filePath, context.file?.fileType, context.file?.fileContent]);
```

tiagobento replied:
Try this:

```
 const file = useMemo(() => ({
      fileName: context.file?.filePath ?? "",
      editorType: context.file?.fileType as EditorType,
      getFileContents: () => Promise.resolve(context.file?.fileContent ?? ""),
      isReadOnly: false
}), [context.file?.filePath, context.file?.fileType, context.file?.fileContent]);
```

manstis commented:
I was using VSCode's default formatting; I've now changed to use prettifier's settings.

I'll fix the `yarn.lock` changes (as you suggest) once I've updated tests.

manstis commented on yarn.lock:
```diff
@@ -9070,14 +9070,7 @@ mkdirp@0.5.1:
-    minimist "^1.2.5"
-
-"mkdirp@>=0.5 0", mkdirp@^0.5.0, mkdirp@^0.5.1, mkdirp@^0.5.4:
+mkdirp@0.x, "mkdirp@>=0.5 0", mkdirp@^0.5.0, mkdirp@^0.5.1, mkdirp@^0.5.4:
```
This is the result of updating `yarn.lock` the way you described.

I replaced my local with the latest in github then ran `yarn install`.

tiagobento commented:
@jomarko Assigning you as the PR features are complete, we're just finishing the tests.

danielezonca commented on packages/desktop/src/webview/App.tsx:
```diff
@@ -48,15 +47,13 @@ export function App(props: Props) {
-          scesimPath: "editors/scesim"
+          dmnPath: "gwt-editors/dmn",
+          bpmnPath: "gwt-editors/bpmn",
+          scesimPath: "gwt-editors/scesim"
```
Does this change have any visible impact on user side?
This is for desktop but I see similar changes on online-editor channel tool and I would like to preserve final URL (user could have a bookmark or saved path that will not work anymore)

manstis replied:
@danielezonca These URLs are _internal_ i.e. used as the `src` in the `IFRAME` containing the editor. The public URLs for the editor remain unchanged and continue to be defined by the `Router`. @tiagobento Please correct me if I'm wrong!

danielezonca replied:
Ok thanks for the explanation
If they are internal I'm fine with that :)

manstis commented:
@tiagobento @jomarko @ljmotta @caponetto This PR is ready for merge....

Just need some approvals; then the PRs this blocks can be updated.

Thank-you for your patience.

manstis commented:
@tiagobento My last question to you...

`EmbeddedEditor` [exposes](https://github.com/kiegroup/kogito-tooling/blob/1a0e8f6beb92c605622baec2beb0614378f2d06f/packages/embedded-editor/src/embedded/EmbeddedEditor.tsx#L43) undo/redo to consumers; however when fixing the unit tests handling of undo/redo seems to be entirely within the inner envelope; delegating to `StateControl`. I therefore wonder whether it best to remove the properties from `EmbeddedEditor`. If we want to expose undo/redo through `EmbeddedEditor` I suspect it'll best be the subject of a new JIRA/PR. WDYT?

tiagobento commented:
@manstis Hi! Let's leave that separate. There will be some work going on to implement undo/redo on channel that are not VS Code soon (which will all use the `EmbeddedEditor`), so this will be addressed.

manstis commented:
Jenkins please retest this.

manstis commented:
@tiagobento IDK if the above phrase works on this repo?

ljmotta commented:
@manstis Hi! (I know that your question wasn't for me, but I think I can answer that :slightly_smiling_face:) The CI on this repo is handled by the GitHub Actions (so the phrase you used is not going to work). If you want to re-run the CI on this branch, I believe that have a few ways to do that:  push/merge to this branch, or trigger it manually (just in case of **failure** someone with the correct permissions over the repo, can re-run the entire workflow).

tiagobento commented:
@manstis @tiagobento Just re-triggered the build.

tiagobento commented on packages/desktop/package.json:
```diff
@@ -11,10 +11,11 @@
-    "@kogito-tooling/microeditor-envelope-protocol": "0.4.3",
+    "@kogito-tooling/embedded-editor": "0.4.3",
     "@kogito-tooling/patternfly-base": "0.4.3",
+    "@patternfly/patternfly": "2.40.2",
+    "@patternfly/react-core": "3.90.0",
+    "@patternfly/react-icons": "3.14.7",
```
@manstis This is not necessary anymore since we get patternfly transitively from patternfly-base

manstis replied:
Removed.

ljmotta commented on packages/desktop/package.json:
```diff
@@ -11,10 +11,11 @@
     "@kogito-tooling/patternfly-base": "0.4.3",
+    "@patternfly/patternfly": "2.40.2",
+    "@patternfly/react-core": "3.90.0",
+    "@patternfly/react-icons": "3.14.7",
```
Hi there! The Patternfly dependencies are managed by the `@kogito-tooling/patternfly-base`, so isn't supposed to have any direct `@patternfly` dependencies on any channel.

manstis commented on packages/embedded-editor/src/envelope/envelope.ts:
```diff
@@ -30,5 +30,9 @@ MicroEditorEnvelope.init({
-  editorContext: { channel: ChannelType.ONLINE, operatingSystem: getOperatingSystem() }
+  editorContext: { channel: getChannelType(), operatingSystem: getOperatingSystem() }
 });
+
```
I'd copied and pasted the content of the `online-editor` _envelope` when rebasing; so had to restore my change.

manstis commented:
@tiagobento Updated following successful merge of your keyboard shortcuts PR :-)

ederign commented:
@jomarko when you have a chance, could you please take a look at this?

jomarko commented:
Manually checked channels:
- [ ] chrome extension

jomarko commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -46,9 +42,13 @@ import {
   ToolbarItem
 } from "@patternfly/react-core";
```
In one of changed files I saw massive removing of similar imports. So just double checking if here is fine to keep them.

manstis replied:
Yes, they appear to still be used by the `HomePage` screen.

manstis replied:
I see [here](https://github.com/kiegroup/kogito-tooling/pull/142/files#diff-87c00c5d80fce81bc8f0044518fdc542R18), which is where you may have seen it.. The imports were reformatted onto a single line by the code-formatter we use (possibly as a result of one or two being removed).. the formatter appears to like splitting/combining lines where their length gets too long.

jomarko commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -227,40 +229,42 @@ export function HomePage(props: Props) {
   );
 
   const createEmptyBpmnFile = useCallback(() => {
-    createEmptyFile("bpmn");
```
just double checking. not sure how deep we want to go. If I `grep` `HomePage.tsx` there are still also "bpmn" "dmn" string occurrences. For exmaple at place where we say what extension of file is accepted. Is that ok?

manstis replied:
I searched `HomePage.tsx` for both `"dmn"` and `"bpmn"` and no occurrences were found.

Can you please advise _where_ you found the occurrences?

manstis commented on packages/chrome-extension/src/app/components/common/KogitoEditorIframe.tsx:
```diff
@@ -136,38 +68,42 @@ const RefForwardingKogitoEditorIframe: React.RefForwardingComponent<IsolatedEdit
+    }
+
+    //keep line breaks
+    const content = editorContent.content.split("\n").join("\\n");
```
Erm, IDK, to be honest. I just rearranged the [existing](https://github.com/kiegroup/kogito-tooling/pull/142/files#diff-3d419b7eeda83b2766633ac29849fb35L71-L72) code to re-use the `embedded-editor`. It looks to have been added [here](https://github.com/kiegroup/kogito-tooling/pull/34/files#r342520114) and possibly the same question was asked at the time.

jomarko replied:
quality engineers .... :)

tiagobento replied:
@jomarko Hi guys!

**tl;dr**

We have to append a script tag with the whole XML hard-coded so that we can use the native text editor that GitHub has.

---
So here's me again with a bit of history and context about our code 😄 

To make sure that the "Commit" button works inside GitHub, we have to "integrate" with the existing CodeMirror instance that's running on GitHub's page. That would be very easy if, as extension developers, we had access to the `window` object of the containing page. Of course that's not the case :)

I think that, for security reasons Chrome, gives the extension code a separate window object, making it not possible for us to access the global CodeMirror classs/function that GtiHub provides. Here's where the workaround begins.

The *only* way we found to use the CodeMirror instance was by using GitHub's window object, so we dynamically create a script tag and fill it with code that counts on having a CodeMirror object globally available. This script tag is then appended to the original document and magically, we have access to GitHub's window object. To make sure that the code is properly formatted, we have to escape every line break so that we don't simple add a script with line breaks, but create a script tag containing a string that has line breaks in it.

Hope that sheds some light onto it :)

manstis commented on packages/desktop/src/webview/App.tsx:
```diff
@@ -48,15 +47,13 @@ export function App(props: Props) {
-          bpmnPath: "editors/bpmn",
-          dmnPath: "editors/dmn",
-          scesimPath: "editors/scesim"
+          dmnPath: "gwt-editors/dmn",
+          bpmnPath: "gwt-editors/bpmn",
+          scesimPath: "gwt-editors/scesim"
```
Removed import. 

An earlier [commit](https://github.com/kiegroup/kogito-tooling/pull/142/commits/9f07fd4fae8421347f7fa717ae760340f25be5a3) in this PR used `EditorType` values but @tiagobento asked for it to be reverted to the literals.

manstis commented on packages/embedded-editor/jest/jest-env-setup.js:
```diff
@@ -0,0 +1,20 @@
+/*
+ *  Copyright 2019 Red Hat, Inc. and/or its affiliates.
```
Changed.

manstis commented on packages/embedded-editor/src/__tests__/envelope/envelope.test.ts:
```diff
@@ -0,0 +1,74 @@
+            return {
+                attributes: {
+                    getNamedItem: (key: string) => {
+                        return { value: key === "data-envelope-channel" ? channelType : undefined };
```
Your observation is 100% correct. The tests just check that the type of channel is correctly extracted from the DOM by reading the value of the `data-envelope-channel` attribute on the enclosing `IFRAME`. This code mocks `getNamedItem(..)` returning the test `channelType` or `undefined` for all other attributes as they're not important for the test.

manstis commented on packages/embedded-editor/src/common/EditorTypes.ts:
```diff
@@ -14,12 +14,8 @@
-export const EMPTY_FILE = {
-  fileName: "new-file",
-  getFileContents: () => Promise.resolve("")
+export enum EditorType {
+  DMN = "dmn",
+  BPMN = "bpmn",
+  SCESIM = "scesim"
```
You've lost me.. this file **is** in the `embedded-editor` module (it was moved from `online-editor` to `embedded-editor` and changed).

jomarko replied:
sorry for confusion @manstis

manstis commented on packages/embedded-editor/src/__tests__/common/File.test.ts:
```diff
@@ -0,0 +1,60 @@
+            });
+    });
+
+describe("File::newFile",
+    () => {
+        test("constructor",
+            async () => {
+                const file: File = newFile(EditorType.DMN);
+                expect(file.fileName).toEqual("new-file");
+                expect(file.editorType).toEqual(EditorType.DMN);
+                expect(file.isReadOnly).toBeFalsy();
+
+                await file.getFileContents().then(data => expect(data).toBe(""));
+            });
+    });
```
Added a new test with content.

manstis commented on packages/chrome-extension/package.json:
```diff
@@ -32,4 +31,4 @@
     ]
   }
-}
+}
```
Added new line. IDK _how_ it went missing but I'll keep an eye on similar in the future.

manstis commented on yarn.lock:
```diff
@@ -13594,4 +13594,4 @@ zip-webpack-plugin@3.0.0:
   dependencies:
     webpack-sources "^1.1.0"
-    yazl "^2.4.3"
+    yazl "^2.4.3"
```
New line added.

manstis commented on packages/embedded-editor/src/__tests__/embedded/EmbeddedViewer.test.tsx:
```diff
@@ -0,0 +1,94 @@
+  await delay(0); //waits til next event loop iteration
+}
+
+describe("EmbeddedViewer", () => {
```
I don't have different tests for different channels however I've updated the "defaults" test for both `EmbeddedEditor` and `EmbeddedViewer` to assert on more values and revised the test suite description to include `::ONLINE`. There are different tests passing different channels in `envelope.test.ts`.

manstis replied:
Corresponding `.snap` files were regenerated as the "defaults" tests now render the component.

manstis commented on packages/embedded-editor/src/__tests__/embedded/EmbeddedViewer.test.tsx:
```diff
@@ -0,0 +1,94 @@
+async function incomingMessage(message: any) {
+  window.postMessage(message, window.location.origin);
+  await delay(0); //waits til next event loop iteration
+}
```
Moved to a test utility class and re-used.

manstis replied:
Documentation added to utilities.

manstis commented:
@jomarko Updates made. Thank-you for your review.

> Would be possible to add typedoc to added public API? Especially in new embedded-editor module. Or does it require some pre-cofiguration. Not just /** .... */ in ts files?

Done to `embedded-editor`

> In some files github reports missing new line in the end of file. Interested if there is soem pros/cons, recommendations about it.

Corrected all I could find.

> my `yarn run build:prod` fails...

It works OK here and on the CI environment. Could I suggest it's something local? After fetching my PR did you `yarn install` it to ensure dependencies were correctly aligned?

jomarko commented:
I still can not build online-editor module. still same error. I tried even `yarn install`, `yarn upgrade` before running `yarn run build:prod`.

Even running `yarn run serve-envelope` in `chrome-extension-pack...` and `yarn start` in online-editor doesn't help.

tiagobento commented:
@jomarko `yarn run init && yarn run build:prod`. Try that :)

ederign commented:
@jomarko when you have a chance, could you please take a look at this?

manstis commented:
@ederign I've (obviously!) merged this and updated the "order" of other PRs (as I noticed @tiagobento had been doing). All I need now is a release of the `kogito-tooling` to `npm`.... when do you plan on your next release? (There's no hurry for me... and it might be an idea to let this PR _settle_ a bit to ensure there's no regressions). Your call. Thank-you.

ederign commented:
+1 for that @manstis. Let's this bake a couple of days.  I'm also fixing a blocker https://issues.redhat.com/browse/KOGITO-2343 .

The only reason for a release in a hurry would be the vscode release. Probably this will happen next week.
